### PR TITLE
Support binary interpolation addition

### PR DESCRIPTION
**Pull request by 333fred**
This adds support for treating binary addition of interpolated strings as if they were a single string, for the purposes of conversions to custom handler types and lowering. I implemented this by delaying binding of binary additions composed entirely of interpolated strings until use, and adjusting all the points that currently expect a string to instead expect either an interpolated string or a binary addition of interpolated strings. @dotnet/roslyn-compiler for review: I'd suggest reviewing commit-by-commit. Commit 1 implements the basic changes just for interpolated strings used as strings, and the second commit adds support for conversions. Closes https://github.com/dotnet/roslyn/issues/54584.

Test plan: https://github.com/dotnet/roslyn/issues/51499

### CODE REVIEW

**Comment by stephentoub** [comment]
Yay. Thanks.

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (chsienki) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:247]: Consider extracting this logic into a local function to be reused below.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs
Comment (chsienki) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs:115]: Consider extracting the body of this if into its own function to make it clearer that we're doing the lowering in a different way.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (chsienki) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:2516]: ?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (chsienki) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Nit: not necesarily as part of this PR, but consider creating a struct for this tuple for ease of reading/understanding, especially when its somewhat divorced from the original context.

**Review by chsienki — APPROVED** [review]

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review.

**Comment by jcouv** [comment]
Looking

**Comment by jcouv** [comment]
> CustomHandler c = " + expression + @";

Do we have tests for various groupings with parens: `($"" + $"") + $""` vs. `$"" + ($"" + $"")`?

---
In reply to: [887677366](https://github.com/dotnet/roslyn/pull/55059#issuecomment-887677366) [](http://example.com/codeflow?ancestors=887677366)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:11710 in 0f97ce1. [](commit_id = 0f97ce173ee8f72b2ec8dc99b1b17708ac74bec8, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:207]: >BoundLiteral [](http://example.com/codeflow?start=35&length=12)

nit: Are those literals necessarily string literals, as the method name "AllInterpolatedStringPartsAreStrings" suggests?

File: src/Compilers/CSharp/Portable/BoundTree/BoundBinaryOperator.UncommonData.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/BoundTree/BoundBinaryOperator.UncommonData.cs:60]: Is it the case that we'll have `InterpolatedStringHandlerData` only after the conversion occurs (ie. when `IsUnconvertedInterpolatedStringAddition` is false)?
Consider asserting any relationship between `IsUnconvertedInterpolatedStringAddition` and `InterpolatedStringHandlerData`.

**Comment by jcouv** [comment]
> CustomHandler c = " + expression + @";

I'm having some trouble getting an overall picture from this PR. If we're converting a suitable unconverted binary op, what are the desired type and converted type for the binary op and for the interpolated string operands? I don't have a clear idea what I should expect.
Similarly, what is the binary operator symbol we'll surface in the semantic model?
Do we have tests for those?

---
In reply to: [887718786](https://github.com/dotnet/roslyn/pull/55059#issuecomment-887718786) [](http://example.com/codeflow?ancestors=887718786,887677366)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:11710 in 0f97ce1. [](commit_id = 0f97ce173ee8f72b2ec8dc99b1b17708ac74bec8, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:327]: Why are we building a `BinaryOperatorSignature` instance, only to use the `Method` (which is `null`) and `ConstrainedToTypeOpt` (also `null`?)?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:385]: >RebindSimpleBinaryOperatorAsConverted [](http://example.com/codeflow?start=33&length=37)

nit: I don't get the "Simple" part. This is specifically a case of a non-simple binary operator (it has unconverted interpolations). 
Maybe `BindUnconvertedInterpolatedStringAddition`? (to follow naming convention of `BindUnconvertedInterpolatedStringToString` above and `IsUnconvertedInterpolatedStringAddition`)

File: src/Compilers/CSharp/Portable/BoundTree/BoundBinaryOperator.UncommonData.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/BoundTree/BoundBinaryOperator.UncommonData.cs:23]: >WithInterpolatedStringHandlerData [](http://example.com/codeflow?start=39&length=33)

There is no instance to "With". Naming should probably align with `UnconvertedInterpolatedStringAddition` above.

File: src/Compilers/CSharp/Portable/BoundTree/Constructors.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/BoundTree/Constructors.cs:421]: The only use of the overload is to add `interpolatedStringHandlerData` to a `BoundBinaryOperator`. Consider adding `Update(InterpolatedStringHandlerData)` instead.

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:318]: Are we losing the constant value of the `BoundBinaryOperator`, and possibly other uncommon data, in this update?

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs:114]: Can we assert that `IsUnconvertedInterpolatedStringAddition` is false? (unconverted nodes don't survive initial binding)

Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs:114]: Aren't we putting a conversion on such binary operators? I would have expected the handling in `LocalRewriter.VisitConversion` would have been sufficient (it handled a conversion on top of a `BoundBinaryOperator` already)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs:131]: Is it possible to encounter a binary op node with `InterpolatedStringHandlerData` aside from the top node? I couldn't find any test in `InterpolationTests` affected by this `if` block.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:2256]: I'm surprised the `AbstractFlowPass` only has to deal with `InterpolatedStringHandlerData` in `VisitBinaryOperator`. Would `VisitConversion` or `VisitCall` (or `VisitArguments...`) need some handling too, like `LocalRewriter`?

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:2411]: Do we have a test that is affected by this change?

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:2548]: We need to release the instance back to the pool

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:2545]: Any chance we could factor some of this logic with `VisitInterpolatedStringBase`? They should be very similar (except that we gather more parts in case of binary op).

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:4037]: It looks lik we only set a nullability result for the top binary op. Don't we need to set a nullability result on every child binary op as well?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:63666]: Could we also test the following with nullability?
- `$"..." + null`
- longer string additions (nested binary ops)

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:234]: nit: `&&=`?

**Review by jcouv — COMMENTED** [review]
Done with review (iteration 3). Haven't yet much looked at tests

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:247]: There's a bunch of methods that visit the binary ops to get to the interpolated strings. Maybe a visitor helper method would make sense? `BoundBinaryOperator.VisitUnconvertedInterpolatedStrings(...)`?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:385]: `Simple` is an existing term. See `BindSimpleBinaryOperator` in `Binder_Operators.cs`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:63666]: The former is not an interpolated string composed entirely of interpolated strings, and is not subject to a handler conversion. I can add more nested ops.

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:234]: This is the second or third time you've asked for this operator. Perhaps you should consider seeing if LDM wants to reconsider the `Likely Never` milestone on [https://github.com/dotnet/csharplang/issues/1718](https://github.com/dotnet/csharplang/issues/1718) 😉

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:318]: Yes, and that is intentional. This binary operator is now being used as part of an interpolated string builder pattern, and the constant value will not be used.

**Comment by 333fred** [comment]
> CustomHandler c = " + expression + @";

> what are the desired type and converted type for the binary op and for the interpolated string operands? I don't have a clear idea what I should expect.

The of the operator and all interpolated string components is `string`, with a converted type of the destination type. This is verified by `VerifyInterpolatedStringExpression`. I can add binary operator symbol validation to that method as well.

---
In reply to: [887936166](https://github.com/dotnet/roslyn/pull/55059#issuecomment-887936166) [](http://example.com/codeflow?ancestors=887936166,887677366)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:11710 in 0f97ce1. [](commit_id = 0f97ce173ee8f72b2ec8dc99b1b17708ac74bec8, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:234]: Oops. This one keeps tripping me up. I even thought I'd checked on sharplab yesterday. :-P

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:4037]: We set one for every child binary op. That's what this method is responsible for.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs:131]: I can add a test for this, but yes it's possible. Consider a series of strings added to an object: the strings will be combined with the default builder, then the standard string + object code will be run.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs:114]: We do. See the change in `BoundExpression.cs` to `NeedsToBeConverted`.

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:207]: Yes, they are.

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:318]: Makes sense.
Let's include tests for `GetConstantValue` on interpolated strings (can have constant value) and binary op (will be null).
Should we also test some `const` scenarios? `const string s = ...`

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:318]: > Should we also test some const scenarios? const string s = ...

I don't think it's worth it. Actually `const` strings aren't lowered using this pattern.

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:247]: I don't really see any particular logic that can be easily shared between the various visits. This location, for example, wants to collect the parts then rebind the interpolated strings, bailing out if it encounters bad data. The abstract flow pass needs to visit every piece. Local rewriting needs to collect the pieces then replace the original operator with a single BoundSequence. IMO, they're all too different to try and make a single helper for the logic.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:2516]: Must have accidentally hit `p` somewhere.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Sure. IOperation changes will come after this, so I can do this in the next PR

**Comment by 333fred** [comment]
> CustomHandler c = " + expression + @";

Deconstruction and tuples appear to just work, actually, so I'll add a couple of tests to cover.

---
In reply to: [888491370](https://github.com/dotnet/roslyn/pull/55059#issuecomment-888491370) [](http://example.com/codeflow?ancestors=888491370,887936166,887677366)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:11710 in 0f97ce1. [](commit_id = 0f97ce173ee8f72b2ec8dc99b1b17708ac74bec8, deletion_comment = False)

**Comment by 333fred** [comment]
@chsienki @jcouv thanks for the feedback. It should be addressed in the latest revision. I'll push another commit later tonight adding the expression tree exception to this PR as well, it's on my home machine and I forgot to push it yesterday.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:2502]: That worked out nicely. Thanks!

**Comment by jcouv** [comment]
>         }

Glad that deconstruction just worked out.
Not blocking this PR, but I was hoping for some explicit tests of `GetConstantValue` to capture the behavior you described (constant value of interpolated strings vs. that of binary ops in different conversion scenarios).

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:12045 in cbac96b. [](commit_id = cbac96bb2d6d1f214eded426ed6e0489a4269bdc, deletion_comment = False)

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:2368]: What was the resolution?

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs:114]: Yeah, I think the problem is more general than what I thought. We should assert that when we get to lowering all nodes have `NeedsToBeConverted` as false. This is not something specific to this PR.

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 5). One question left unaddressed

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:2368]: This is the same test that addressed the nested binary operator question: `InterpolatedStringsAddedUnderObjectAddition`. That will hit all the codepaths that need to break out of a nested binary operation, both in flow analysis and rewriting.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:2368]: I've run the interpolated and nullable tests without this change and none was affected. I'm fine to move ahead (the change seems correct) but please try to find a way to test it.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 6). Please follow-up on a test that is affected (remaining open issue)

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:2368]: Added a definite assignment test that will fail on this codepath.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (chsienki) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:9788]: Nit: can we get rid of the commented code? either enter the function and return node, or remove.

Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:9788]: I intentionally left this in to leave a place for the issue comment.

**Comment by jcouv** [comment]
Will this target 17.0p3? Branch should change if so

**Review by chsienki — APPROVED** [review]

**Pull request merged**
